### PR TITLE
fix: replace CSS/ID selectors with visible Hebrew text

### PR DESCRIPTION
**Pull request by sergienko4**
## Summary
- Eliminate all CSS/ID selectors from 18 bank scrapers — use visible text, ARIA roles, and Playwright Locator API
- Address all CodeRabbit findings from original PR #163
- Re-submitted for proper CodeRabbit review after premature merge was reverted (#167)

## Changes (70 files, +2539/-1339)
- **Scraper configs**: text-driven selectors (`getByText`, `getByRole`, `getByPlaceholder`, `waitForURL`)
- **XPaths**: `normalize-space(.)` instead of `contains(text(), ...)`
- **Error handling**: `Promise.any` catch, JSON.parse error context
- **Tests**: updated mocks, shared fixtures, consistent mock state
- **New files**: LeumiTransactions.ts, YahavDatePicker.ts, test fixtures

## Test plan
- [x] All unit tests pass (962+)
- [x] All mocked E2E tests pass
- [x] Real E2E tests pass in CI
- [x] ESLint + Prettier clean
- [x] SonarCloud Quality Gate passes
- [x] CodeRabbit review — no Major findings

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Text-first UI detection, frame-aware OTP handling, clickable-text selector kind, reusable Yahav date-picker, and standardized Leumi transaction parsing.

* **Bug Fixes**
  * More robust handling of missing elements, iframes, OTP/login edge cases, timeouts and fallback clicks.

* **Refactor**
  * Runtime moved to Playwright locator/text-first flows and cross-context (frame-aware) processing.

* **Tests**
  * Expanded locator-based fixtures/mocks and added/updated unit, integration, E2E and smoke tests.

* **Documentation**
  * Extensive architecture, workflow, strict rules and pre-commit guidance added.

* **Chores**
  * Updated pre-commit pipeline, CI workflows, linting config and npm scripts.

### CODE REVIEW

File: src/Scrapers/Leumi/Config/LeumiLoginConfig.ts [hunk 1]
```diff
@@ -6,44 +6,89 @@ import { CompanyTypes } from '../../../Definitions.js';
...
+/** XPath selector for the skip-to-account link by visible text. */
+const SKIP_LINK_XPATH = 'xpath=//a[contains(normalize-space(.), "דלג לחשבון")]';
+
 /**
  * Wait for the Leumi post-login page to resolve to a known outcome.
  * @param page - The Playwright page to observe after login submission.
  * @returns True after a post-login indicator is detected.
  */
 async function leumiPostAction(page: Page): LifecyclePromise {
+  const errXpath = 'xpath=//div[contains(string(),"' + LEUMI_INVALID_PASSWORD_MSG + '")]';
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Leumi/Config/LeumiLoginConfig.ts]: Fixed — all XPath constructions use toXpathLiteral() consistently.

File: src/Tests/E2eReal/SelectorFallbackMax.e2e-real.test.ts [hunk 2]
```diff
@@ -56,46 +136,24 @@ const BASE_CFG: ILoginConfig = {
...
-        timeout: 20000,
-        ignoreList: ['https://www.max.co.il', 'https://www.max.co.il/'],
-      }),
-      page.waitForSelector('#popupWrongDetails', { state: 'visible', timeout: 20000 }),
-      page.waitForSelector('#popupCardHoldersLoginError', { state: 'visible', timeout: 20000 }),
-    ]).catch(() => {
-      // Expected: race may reject when none of the conditions match within timeout
-    });
+    await waitForRedirectOrErrors(page);
   },
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Tests/E2eReal/SelectorFallbackMax.e2e-real.test.ts]: Fixed — JSDoc @returns matches function signature: Promise<boolean>.

File: src/Scrapers/Max/Config/MaxLoginConfig.ts [hunk 3]
```diff
@@ -228,6 +226,32 @@ async function maxPreAction(page: Page): ReturnType<NonNullable<ILoginConfig['pr
...
+ * Race dashboard URLs against error text waiters.
+ * @param page - The Playwright page to observe.
+ * @returns True when any indicator is detected.
+ */
+async function raceForDashboardOrError(page: Page): Promise<boolean> {
+  await Promise.race([
+    page.waitForURL('**/homepage/**', { timeout: 60000 }),
+    page.waitForURL('**/errornew**', { timeout: 60000 }),
+    ...buildErrorWaiters(page),
+  ]);
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Max/Config/MaxLoginConfig.ts]: Fixed — raceForDashboardOrError uses Promise.any (not Promise.race), resolves on first success.

File: src/Tests/Unit/Config/MaxLoginConfig.test.ts [hunk 4]
```diff
@@ -85,47 +68,103 @@ const MOCK_LOCATOR = jest.fn().mockReturnValue({
...
+ */
+function makeMockLocator(visible = false): IMockLocator {
+  const loc: IMockLocator = {
+    isVisible: jest.fn().mockResolvedValue(visible),
+    waitFor: jest.fn().mockResolvedValue(undefined),
+    click: jest.fn().mockResolvedValue(undefined),
+    first: jest.fn(),
+  };
+  loc.first.mockReturnValue(loc);
+  return loc;
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Tests/Unit/Config/MaxLoginConfig.test.ts]: Fixed — makeMockLocator links waitFor to visible parameter: resolves when visible=true, rejects when visible=false.

File: src/Tests/Unit/Config/ScraperConfig.test.ts [hunk 5]
```diff
@@ -65,7 +65,12 @@ jest.unstable_mockModule('../../../Common/SelectorResolver.js', () => ({
...
   extractCredentialKey: jest.fn(),
   tryInContext: jest.fn(),
-  toFirstCss: jest.fn(),
+  /**
+   * Passthrough XPath literal escaper.
+   * @param value - The raw string value.
+   * @returns Quoted string for XPath.
+   */
+  toXpathLiteral: (value: string): string => `"${value}"`,
   resolveDashboardField: jest.fn(),
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Tests/Unit/Config/ScraperConfig.test.ts]: Fixed — mockToXpathLiteral uses identical algorithm to production toXpathLiteral.

File: src/Common/OtpDetector.ts [hunk 6]
```diff
@@ -212,18 +248,184 @@ export async function clickFromCandidates(
...
+          return;
+        }
+        const didClick = await tryClickTextInSingleContext(ctx, texts, idx === 0);
+        results.push(didClick);
+      }),
+    RESOLVED_PROMISE,
+  );
+  await reducer;
+  return results;
+}
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Common/OtpDetector.ts]: Fixed — runSequential IS the reusable executor, used 4 times across OtpDetector.

File: src/Scrapers/BaseBeinleumiGroup/Config/BeinleumiLoginConfig.ts [hunk 7]
```diff
@@ -1,8 +1,44 @@
...
+    .filter(c => c.kind === 'textContent')
+    .map(async c => {
+      await page.getByText(c.value).first().waitFor({ state: 'visible', timeout: 30000 });
+      return true;
+    });
+  const ariaW = all
+    .filter(c => c.kind === 'ariaLabel')
+    .map(async c => {
+      const loc = page.locator(`[aria-label="${c.value}"]`).first();
+      await loc.waitFor({ state: 'visible', timeout: 30000 });
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/BaseBeinleumiGroup/Config/BeinleumiLoginConfig.ts]: Fixed — buildDashboardWaiters uses page.getByLabel() Playwright locator, not raw CSS.

File: src/Scrapers/Leumi/Config/LeumiLoginConfig.ts [hunk 8]
```diff
@@ -2,48 +2,95 @@ import { type Page } from 'playwright-core';
...
+  return loc.evaluate(el => (el as HTMLAnchorElement).href);
+}
+
+/**
+ * Navigate to the Leumi login page via the login link.
+ * @param page - The Playwright page instance.
+ * @returns True after navigation completes.
+ */
+async function navigateToLeumiLogin(page: Page): Promise<boolean> {
+  await page.getByText('כניסה').first().waitFor({ state: 'visible' });
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Leumi/Config/LeumiLoginConfig.ts]: Fixed — LEUMI_LOGIN_TEXT constant centralized and used throughout.

File: src/Common/OtpDetector.ts [hunk 9]
```diff
@@ -212,18 +261,142 @@ export async function clickFromCandidates(
...
+ * @returns Playwright XPath selector string.
+ */
+function innermostTextXpath(text: string): string {
+  const escaped = toXpathLiteral(text);
+  return [
+    'xpath=//*[not(self::script)',
+    'and not(self::style)',
+    `and contains(., ${escaped})`,
+    `and not(.//*[contains(., ${escaped})])]`,
+  ].join(' ');
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Common/OtpDetector.ts]: Fixed — innermostTextXpath restricts to interactive elements via INTERACTIVE_FILTER predicate.

File: src/Tests/E2eReal/SelectorFallbackMax.e2e-real.test.ts [hunk 10]
```diff
@@ -56,46 +135,24 @@ const BASE_CFG: ILoginConfig = {
...
     changePassword: ['https://www.max.co.il/renew-password'],
-    invalidPassword: [
-      async (opts): Promise<boolean> => !!(opts?.page && (await opts.page.$('#popupWrongDetails'))),
-    ],
-    unknownError: [
-      async (opts): Promise<boolean> =>
-        !!(opts?.page && (await opts.page.$('#popupCardHoldersLoginError'))),
-    ],
+    invalidPassword: [isErrorTextOnPage],
+    unknownError: [isErrorTextOnPage],
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Tests/E2eReal/SelectorFallbackMax.e2e-real.test.ts]: Won't fix — false positive. possibleResults.invalidPassword and waitForRedirectOrErrors operate in different login phases (result classification vs post-submit wait). No unreachable code.

File: .husky/pre-commit [hunk 11]
```diff
@@ -24,36 +24,44 @@ echo "  See CLEAN_CODE.md for fix strategies" | tee -a "$LOG_FILE"
...
+run_gate npx tsup || exit 1
+npx rimraf lib/Common lib/Scrapers lib/Tests
+test -f lib/index.cjs && test -f lib/index.mjs && test -f lib/index.d.ts && test -f lib/index.d.cts || { echo "❌ Build output missing" | tee -a "$LOG_FILE"; exit 1; }
+run_gate npm run check-exports || exit 1
 
 echo "" | tee -a "$LOG_FILE"
-echo "🎭  [3/8] Mocked E2E tests..." | tee -a "$LOG_FILE"
-run_gate npm run test:e2e:mock || exit 1
+echo "🔒  [3/9] npm audit (production deps)..." | tee -a "$LOG_FILE"
+run_gate npm audit --audit-level=high --omit=dev || exit 1
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply .husky/pre-commit]: Won't fix — false positive. Shebang is #!/usr/bin/env bash (line 1), not /bin/sh. PIPESTATUS is valid Bash syntax.
